lib/vrf.c

```diff
@@ -417,39 +466,74 @@ void vrf_terminate(void)
 /* Create a socket for the VRF. */
 int vrf_socket(int domain, int type, int protocol, vrf_id_t vrf_id)
 {
-	int ret = -1;
+	int ret, save_errno;
+	struct vrf *vrf = vrf_lookup_by_id(vrf_id);
 
+	vrf_switch_to_netns(vrf);
```

Comment:
what kind of errors are you thinking of ? some panic ? or some zlog_err message ?

Reply:
The `setns()` syscall can fail for a variety of reasons. Your code for example is susceptible to race conditions since a netns can be deleted right before you're trying to `setns()` to it. The `setns()` man page mentions several other error possibilities.

What I'd do is to add a `zlog_err()` message in `ns_switch_to_netns()` when `setns()` fails. Both `ns_switch_to_netns()` and `vrf_switch_to_netns()` already return negative values when something goes wrong, we shouldn't ignore these values. In this function for example, we should abort the socket creation if `setns()` failed.

Reply:
ok. I will make a proposal in the usage.
- vrf_switch ( back) returns errors only if setns() is called
- errors are zlog_err()
the call is a little bit like it is done when you want to raise or lower priority.